src/bifrost/core.clj

```diff
@@ -43,24 +38,47 @@
                                (:path-params request)
                                (:bifrost-params request))))
 
+(defn ctx->bifrost-request [ctx]
+  (let [request (:request ctx)]
+    (assoc request :bifrost-params (params-map request))))
+
+(def interceptor-xf
+  (map (fn [[response-ch ctx]]
+         [response-ch (ctx->bifrost-request ctx)])))
+
+(defn api-response->ctx
+  [api-response]
+  (let [status (response->http-status api-response)]
+    {:response (-> api-response
+                   (dissoc :status)
+                   ring-resp/response
+                   (ring-resp/status status))}))
+
+(def api-response-xf (map api-response->ctx))
+
 (defn async-interceptor
-  [channel]
-  (interceptor
-   {:enter
-    (fn [ctx]
-      (let [response-channel (async/chan 1)
-            request (:request ctx)]
-        (async/>!! channel [response-channel (params-map request)])
-        (async/go
-          (let [response (async/alt! (async/timeout *response-timeout*) timeout-response
-                                     response-channel ([r] r))
-                status (response->http-status response)]
-            (assoc ctx :response
-                   (-> response
-                       (dissoc :status)
-                       ring-resp/response
-                       (ring-resp/status status)))))))}))
+  ([channel]
+   (async-interceptor channel (gensym)))
+  ([channel response-channel-key]
+   (async-interceptor channel response-channel-key (map identity)))
+  ([channel response-channel-key response-channel-xf]
+   (let [response-channel (async/chan 1 response-channel-xf)]
+     (interceptor/interceptor
+      {:enter
+       (fn [ctx]
+         (async/>!! channel [response-channel ctx])
+         (assoc-in ctx [:response-channels response-channel-key] response-channel))
+       :leave
+       (fn [ctx]
+         (if-let [response (async/alt!!
+                             (async/timeout *response-timeout*) {:response {:status 500 :body "Timeout"}}
+                             response-channel ([r] r))]
+           (merge ctx response)
+           ctx))}))))
 
-(extend-protocol IntoInterceptor
-  clojure.core.async.impl.channels.ManyToManyChannel
-  (-interceptor [ch] (async-interceptor ch)))
```

Comment:
Ah, I see.




What if we just gensym'd (or similar) a keyword for each one if you just put the channel into the route map, and said in the docs that if you want to access the response in your own interceptor, you should use the `interceptor` macro directly?




I'm mostly just curious at this point. The protocol seemed cool so it's sad to let it go. But from a "good library" perspective, I do like things that work one way, consistently. So I'm honestly not sure it's even a good idea to try to keep the protocol.


—
Wes MorganChief Technology Officer

Democracy Works, Inc. (We make TurboVote and Ballot Scout)

+1.720.319.7591

wes@democracy.works

On Thu, Sep 10, 2015 at 9:09 AM, Christopher Shea
<notifications@github.com> wrote:

>>  
>> -(extend-protocol IntoInterceptor
>> -  clojure.core.async.impl.channels.ManyToManyChannel
>> -  (-interceptor [ch] (async-interceptor ch)))
> Sadly, no. The macro works not just on the channel, but also name of the arg itself. Using the macro here would result in all of the channels getting assoced into the context at `[:response-channels :ch]` (so only the last interceptor would get its channel on the context).
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/democracyworks/bifrost/pull/1/files#r39171152